util/bloom_test.cc

```diff
@@ -601,6 +603,255 @@ TEST_P(FullBloomTest, OptimizeForMemory) {
+  constexpr std::size_t kCacheCapacity = 8 * 1024 * 1024;
+  std::size_t memory_reserved = 0;
+  std::size_t num_dummy_entries = 0;
+
+  BlockBasedTableOptions table_options;
+  LRUCacheOptions lo;
+  lo.capacity = kCacheCapacity;
+  lo.num_shard_bits = 0;  // 2^0 shard
+  lo.strict_capacity_limit = true;
+  std::shared_ptr<Cache> cache(NewLRUCache(lo));
+  table_options.block_cache = cache;
+  table_options.filter_policy.reset(
+      new BloomFilterPolicy(FLAGS_bits_per_key, mode));
+  table_options.reserve_table_builder_memory = true;
+
+  FilterBuildingContext ctx(table_options);
+
+  ASSERT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+  std::unique_ptr<FilterBitsBuilder> filter_bits_builder(
+      table_options.filter_policy->GetBuilderWithContext(ctx));
+
+  const std::size_t num_entries = 10000;
+  char key_buffer[sizeof(int)];
+  for (std::size_t i = 0; i < num_entries; ++i) {
+    filter_bits_builder->AddKey(Key(static_cast<int>(i), key_buffer));
+  }
+  // 0xffffffffU is max size supported by implementation for buf length
+  std::unique_ptr<const char[]> buf(new char[0xffffffffU]);
+  Slice filter = filter_bits_builder->Finish(&buf);
+
+  if (can_reserve_memory) {
+    // Memory reserved now: mutable_buffer_'s size (approximated by
+    // filter.size())
+    memory_reserved = filter.size();
+  } else {
+    memory_reserved = 0;
+  }
+  num_dummy_entries = static_cast<std::size_t>(
+      std::ceil(memory_reserved / static_cast<double>(kSizeDummyEntry)));
+  EXPECT_GE(cache->GetPinnedUsage(), num_dummy_entries * kSizeDummyEntry);
+  EXPECT_LT(cache->GetPinnedUsage(), (num_dummy_entries + 1) * kSizeDummyEntry);
+
+  filter_bits_builder.reset();
+  EXPECT_EQ(cache->GetPinnedUsage(), 0 * kSizeDummyEntry);
+}
+
+TEST_P(FullBloomTest, RibbonFilterFallBackOnLargeBanding) {
+  const BloomFilterPolicy::Mode mode = GetParam();
+  if (mode != BloomFilterPolicy::Mode::kStandard128Ribbon) {
+    return;
+  }
+
+  constexpr std::size_t kSizeDummyEntry = 256 * 1024;
+  constexpr std::size_t kCacheCapacity = 8 * 1024 * 1024;
+  std::size_t memory_reserved = 0;
+  std::size_t num_dummy_entries = 0;
+
+  // Number 500,000 is carefully chosen so that memory reserved before banding
+  // won't cause a cache full under strict capacity limit but will after.
+  // Therefore it will trigger a fallback due to banding portion being too large
+  // during FilterBitsBuilder::Finish().
+  //
+  // Calculation:
+  // Before banding: hash_entries_.size() * hash_entries_[0] ≈ 500,000 * 8 =
+  // 4,000,000 > 3MB Banding'size ≈ 10,437,168 > 9MB After banding: 3MB + 9MB >
+  // 8MB
+  //
+  // Number 50,000 is a small number so there will not be any cache full hence
+  // fallback.
+  //
+  // Calculation:
+  // Before banding: hash_entries_.size() * hash_entries_[0] ≈ 50,000 * 8 =
+  // 400,000 < 0.4MB Banding'size ≈ 1,034,288 < 2MB After banding: 0.4 MB + 2MB
+  // < 8MB
+  for (std::size_t num_entries : {500000, 50000}) {
```

Comment:
It should not be necessary to use large, expensive constructions in unit tests. You should be able to scale the parameters down to something super cheap and fast. (We are not introducing new reasons to be concerned with e.g. value overflow.)